akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala

```diff
@@ -0,0 +1,111 @@
+ * Uses the seed as a key to encrypt a 128-bit counter using AES(Rijndael).
+ *
+ * Keys larger than 128-bit for the AES cipher require
+ * the inconvenience of installing the unlimited strength cryptography policy
+ * files for the Java platform.  Larger keys may be used (192 or 256 bits) but if the
+ * cryptography policy files are not installed, a
+ * java.security.GeneralSecurityException will be thrown.
+ *
+ * NOTE: THIS CLASS IS NOT SERIALIZABLE
+ */
+
+class AESCounterBuiltinCTRRNG(val seed: Array[Byte]) extends Random {
+  private val COUNTER_SIZE_BYTES = 16
+  private val BITWISE_BYTE_TO_INT = 0x000000FF
+  private val RESEEDING_THRESHOLD = 1000000 // million
+  private val RESEEDING_DEADLINE = 140737488355328L // 2^47 safe as per SP800-90
+
+  private final val SOURCE = new SecureRandom
+
+  // mutable state below, concurrent accesses need lock
+  private val lock = new ReentrantLock
+  private val counter: Array[Byte] = Array.fill[Byte](COUNTER_SIZE_BYTES)(0)
+  private var index: Int = 0
+  private var currentBlock: Array[Byte] = null
+  private var reseedFuture: Future[Array[Byte]] = null
+  private var bitsSinceSeeding: Long = 0
+
+  private val cipher = Cipher.getInstance("AES/CTR/NoPadding")
+  val ivArr = Array.fill[Byte](COUNTER_SIZE_BYTES)(0)
+  ivArr(0) = (ivArr(0) + 1.toByte).toByte
+  private val ivSpec = new IvParameterSpec(ivArr)
+  cipher.init(Cipher.ENCRYPT_MODE, new this.AESKey(seed), ivSpec)
+
+  @Override
+  override protected def next(bits: Int): Int = {
+    try {
+      lock.lock()
+      bitsSinceSeeding += bits
+      if (currentBlock == null || currentBlock.length - index < 4) {
+        try {
+          currentBlock = cipher.doFinal(counter)
+          index = 0
+        } catch {
+          case ex: Exception ⇒ {
+            // Generally Cipher.doFinal() from nextBlock may throw various exceptions.
+            // However this should never happen.  If initialisation succeeds without exceptions
+            // we should be able to proceed indefinitely without exceptions.
+            throw new IllegalStateException("Failed creating next random block.", ex)
+          }
+        }
+      }
+      val result = (BITWISE_BYTE_TO_INT & currentBlock(index + 3)) |
+        ((BITWISE_BYTE_TO_INT & currentBlock(index + 2)) << 8) |
+        ((BITWISE_BYTE_TO_INT & currentBlock(index + 1)) << 16) |
+        ((BITWISE_BYTE_TO_INT & currentBlock(index)) << 24)
+
+      index += 4
+      result >>> (32 - bits)
+    } finally {
+      import scala.concurrent.ExecutionContext.Implicits.global
+      if (bitsSinceSeeding > RESEEDING_THRESHOLD) {
+        if (reseedFuture == null) {
+//          println(s"reseeding started! bits: $bitsSinceSeeding")
+          reseedFuture = Future { SOURCE.generateSeed(32) }
+        }
+        if (bitsSinceSeeding > RESEEDING_DEADLINE) {
+//          println(s"blocking for the seeding result bits: $bitsSinceSeeding")
+          Await.ready(reseedFuture, Duration.Inf)
```

Comment:
The thing inside the future is basically just a read from `/dev/random` (or `/dev/urandom` on BSD or system call on Windows). If it does not return in reasonable timeframe, then probably something is very wrong. But I could imagine a retry logic backed up by an exception after a couple of failed attempts, WDYT?